File: src/DSharp.Mscorlib/Lazy.cs
using System.Runtime.CompilerServices;

namespace System
{
    [ScriptIgnoreGenericArguments]
    public class Lazy<T>
    {
        [ScriptAlias("IsValueCreated")]
        public extern bool IsValueCreated { get; }

        [ScriptAlias("Value")]
        public extern T Value { get; }

        public extern Lazy(Func<T> factory);
    }
}

File: src/DSharp.Script/src/System/Lazy.js
function Lazy(factory)
{
    this.factory = factory;
    this.isValueCreated = false;
    this.value = undefined;
}

var Lazy$ = {
    $get_IsValueCreated: function () {
        return this.isValueCreated;
    },
    $get_Value: function () {
        if (!this.isValueCreated) {
            this.value = this.factory();
            this.isValueCreated = true;
        }
        return this.value;
    }
}